tests/func/utils/test_strict_yaml.py

```diff
@@ -0,0 +1,310 @@
+import os
+
+import pytest
+from ruamel.yaml import __with_libyaml__ as ruamel_clib
+
+from dvc.main import main
+from dvc.ui import ui
+
+DUPLICATE_KEYS = """\
+stages:
+  stage1:
+    cmd: python train.py
+    cmd: python train.py
+"""
+
+DUPLICATE_KEYS_OUTPUT = """\
+'./dvc.yaml' is corrupted.
+
+While constructing a mapping, in line 3, column 5
+  3 │   cmd: python train.py
+
+Found duplicate key "cmd" with value "python train.py" (original value:\
+ "python \ntrain.py"), in line 4, column 5
+  4 │   cmd: python train.py"""
+
+
+MAPPING_VALUES_NOT_ALLOWED = """\
+stages:
+  stage1
+    cmd: python script.py
+"""
+
+MAPPING_VALUES_NOT_ALLOWED_OUTPUT = """\
+'./dvc.yaml' is corrupted.
+
+Mapping values are not allowed {}, in line 3, column 8
+  3 │   cmd: python script.py""".format(
+    "in this context" if ruamel_clib else "here"
+)
+
+
+NO_HYPEN_INDICATOR_IN_BLOCK = """\
+stages:
+  stage1:
+    cmd: python script.py
+    outs:
+      - logs:
+          cache: false
+      metrics:
+"""
+
+NO_HYPEN_INDICATOR_IN_BLOCK_OUTPUT = """\
+'./dvc.yaml' is corrupted.
+
+While parsing a block collection, in line 5, column 7
```

Comment:
I find this a lot easier to understand by looking at the entire `outs:` block.

Reply:
This is a syntax error. So we don't really have a way to go one node above, as we don't even get a semi-parsed dictionary from ruamel. The lines that we show are what we get from the error directly. However, we can show `n` lines before and after that point.

Reply:
For the schema errors, we can traverse upwards, but we don't have a way to know the line ranges, `endpos` in particular. There may be a workaround, but they are a slippery slope.

Reply:
Yeah, personally I'd be in favor of generally including a line or two above and below and highlighting the error line to give more context, but curious if you think this is helpful or too noisy.

Reply:
It may be noisy, as it will be about 5lines of text per error message. With 2 errors on average, it's 12 lines of the message (2*5 lines + 1 main header line + 1 separation line), which is half of `80*24` standard terminal.

I haven't seen other programming languages do this, but this is YAML 😄, so adding context does make sense, although it's better to be precise in error messages than the code snippets. Everything after that message is a best effort.

Reply:
True. There's no obvious right answer, so up to you.

Reply:
At least for code related things, I prefer when linting tools default to displaying the single line and then allow the user to configure displaying `n` lines before after via env vars or command flags